draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -356,869 +524,832 @@ document are to be interpreted as described in {{RFC2119}}.
-
-| Curve  | Inj. Encoding | Random Oracle |
-|--------|---------------|------|
-| P-256 | Simple SWU {{simple-swu}} | FFSTV(SWU) {{ffstv}}
-| P-384 | Icart {{icart}} | FFSTV(Icart) {{ffstv}}
-| Curve25519 | Elligator2 {{elligator2}} | FFSTV(Elligator2) {{ffstv}}
-| Curve448 | Elligator2 {{elligator2}} | FFSTV(Elligator2) {{ffstv}}
+Cryptographic protocols which are proven secure in the random oracle model (ROM)
+often require a hash function that behaves as a random oracle, i.e., that its
+response be uniformly distributed on the set of outputs (uniformity property).
 
-# Utility Functions {#utility}
+Instantiating one of these protocols with an elliptic curve group motivates
+the term "hashing to the curve", i.e. mapping bitstrings to points on the
+elliptic curve. This function can be easily accomplished by using the output of
+a cryptographically secure hash function H as the input of an encoding function
+f.
 
-Algorithms in this document make use of utility functions described below.
+~~~
+hash2curve(alpha) := f(H(alpha))
+~~~
 
-- hash2base(x).
-  This method is parametrized by p and H, where p is the prime order of
-  the base field Fp, and H is a cryptographic hash function which
-  outputs at least floor(log2(p)) + 1 bits.
-  The function first hashes x, converts the result to an integer,
-  and reduces modulo p to give an element of Fp.
-  We provide a more detailed algorithm in {{hashtobase}}.
+On the one hand, this function is not invertible since it is computationally
+intractable to produce an input alpha that maps to hash2curve(alpha) due to H is
+pre-image resistant.
+On the other hand, the uniformity property is not met as the output of an
+encoding is distinguishable from a random distribution. Hence, this function is
+not sufficient to get uniformity, but it can be used as a building block as
+described in {{rom}}.
 
-- CMOV(a, b, c): If c = 1, return a, else return b.
+# Utility Functions {#utility}
 
-  Common software implementations of constant-time selects assume c = 1 or c = 0. CMOV
-  may be implemented by computing the desired selector (0 or 1) by ORing all bits of c
-  together. The end result will be either 0 if all bits of c are zero, or 1 if at least
-  one bit of c is 1.
+Algorithms in this document make use of utility functions described below.
 
-- CTEQ(a, b): Returns a == b.
-  Inputs a and b must be the same length (as bytestrings) and the comparison
-  must be implemented in constant time.
+- is_square(x, q): It returns True whenever the value x is a quadratic residue
+  in a field of order q. Due to Euler's criterion, this function can be
+  calculated in constant time as
+
+  ~~~
+    is_square(x, q) := { True,  if x^((q-1)/2) is 0 or 1;
+                       { False, otherwise.
+  ~~~
+
+  For clarity, it is generally preferred to use this formula directly, and
+  annotate its usage with this definition.
+
+- sqrt(x, q): The sqrt operation is a multi-valued function, i.e. there exist
```

Comment:
The code uses `sq_root` in some places. Can we make this usage consistent?

Reply:
Good catch.